### ISSUE TITLE
peer closed connection in SSL handshake while SSL handshaking to upstream

### ISSUE THREAD

**Issue by zedtux** [rpardini/docker-registry-proxy#18]
I'm facing this SSL handshake error when trying to pull a Kubernetes image:

```
$ docker pull k8s.gcr.io/kube-proxy-amd64:v1.10.4
Error response from daemon: received unexpected HTTP status: 502 Bad Gateway
```

And in this image's logs:

```
2019/08/09 12:44:59 [alert] 105#105: 1024 worker_connections are not enough
2019/08/09 12:44:59 [error] 106#106: *16512 peer closed connection in SSL handshake while SSL handshaking to upstream, client: 127.0.0.1, server: _, request: "GET /v2/ HTTP/1.1", upstream: "https://74.125.140.82:443/v2/", host: "k8s.gcr.io"
- [09/Aug/2019:12:44:59 +0000] "/v2/" 502 173 "HOST:k8s.gcr.io" "PROXY-HOST:k8s.gcr.io" "UPSTREAM:74.125.140.82:443"
2019/08/09 12:45:00 [alert] 105#105: 1024 worker_connections are not enough
2019/08/09 12:45:00 [error] 106#106: *17538 peer closed connection in SSL handshake while SSL handshaking to upstream, client: 127.0.0.1, server: _, request: "GET /v2/kube-proxy-amd64/manifests/v1.10.4 HTTP/1.1", upstream: "https://74.125.140.82:443/v2/kube-proxy-amd64/manifests/v1.10.4", host: "k8s.gcr.io"
- [09/Aug/2019:12:45:00 +0000] "/v2/kube-proxy-amd64/manifests/v1.10.4" 502 173 "HOST:k8s.gcr.io" "PROXY-HOST:k8s.gcr.io" "UPSTREAM:74.125.140.82:443"
```

I'm using Docker Desktop 2.1.0.0, Channel edge which bundle Docker Engine 19.03.1, I created the container with the following command:

```
$ docker run --name docker-registry_proxy \
             --restart=always \
             --publish 3128:3128 \
             --volume /private/var/lib/docker-registry/cache:/docker_mirror_cache \
             --volume /private/var/lib/docker-registry/ca:/ca \
             --env REGISTRIES="k8s.gcr.io gcr.io quay.io" \
             rpardini/docker-registry-proxy:0.2.4
```

_I'm not passing the `AUTH_REGISTRIES` env variable as I only want to cache image pulling, I don't want to pull private images._

To setup the proxy on the Mac I did the following:

1. Download the `ca.crt` file:
  ```
  $ curl http://127.0.0.1:3128/ca.crt > ~/Downloads/ca.crt
  ```
2. I double clicked it, which opened the **Keychain Access** app and I imported it as a system certificate
3. I then selected the certificate from the **Keychain Access** app and selected **Always Trust** from "When using this certificate:".
4. I went to the Docker Desktop > Preferences > Resources > Proxies and I filled in both HTTP and HTTPS with `http://127.0.0.1:3128/` and restarted Docker

Can you please help me in order to get this setup working?

---

Trying to grab some information using curl ...

Here is a curl to `k8s.gcr.io` without the proxy:

```
curl -vv https://k8s.gcr.io/
*   Trying 2a00:1450:400c:c09::52...
* TCP_NODELAY set
*   Trying 173.194.76.82...
* TCP_NODELAY set
* Connected to k8s.gcr.io (173.194.76.82) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-ECDSA-CHACHA20-POLY1305
* ALPN, server accepted to use h2
* Server certificate:
*  subject: C=US; ST=California; L=Mountain View; O=Google LLC; CN=*.gcr.io
*  start date: Jul 29 17:26:57 2019 GMT
*  expire date: Oct 27 17:26:57 2019 GMT
*  subjectAltName: host "k8s.gcr.io" matched cert's "*.gcr.io"
*  issuer: C=US; O=Google Trust Services; CN=GTS CA 1O1
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x7f86b1002600)
> GET / HTTP/2
> Host: k8s.gcr.io
> User-Agent: curl/7.54.0
> Accept: */*
>
* Connection state changed (MAX_CONCURRENT_STREAMS updated)!
< HTTP/2 302
< location: https://cloud.google.com/container-registry/
< content-length: 241
< date: Fri, 09 Aug 2019 13:21:11 GMT
< content-type: text/html; charset=UTF-8
< server: Docker Registry
< x-xss-protection: 0
< x-frame-options: SAMEORIGIN
< alt-svc: quic=":443"; ma=2592000; v="46,43,39"
<
<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>302 Moved</TITLE></HEAD><BODY>
<H1>302 Moved</H1>
The document has moved
<A HREF="https://cloud.google.com/container-registry/">here</A>.
</BODY></HTML>
* Connection #0 to host k8s.gcr.io left intact
```

And here with the proxy:

```
HTTPS_PROXY="http://192.168.178.26:3128/" curl -vv https://k8s.gcr.io/
*   Trying 192.168.178.26...
* TCP_NODELAY set
* Connected to 192.168.178.26 (192.168.178.26) port 3128 (#0)
* Establish HTTP proxy tunnel to k8s.gcr.io:443
> CONNECT k8s.gcr.io:443 HTTP/1.1
> Host: k8s.gcr.io:443
> User-Agent: curl/7.54.0
> Proxy-Connection: Keep-Alive
>
< HTTP/1.1 200 Connection Established
< Proxy-agent: nginx
<
* Proxy replied OK to CONNECT request
* ALPN, offering h2
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: C=NL; ST=Noord Holland; L=Amsterdam; O=ME; OU=IT; CN=DockerMirrorBox Web Cert d9add2a9f015 2019.08.09 13:09
*  start date: Aug  9 13:09:16 2019 GMT
*  expire date: Aug  8 13:09:16 2020 GMT
*  subjectAltName: host "k8s.gcr.io" matched cert's "k8s.gcr.io"
*  issuer: C=NL; ST=Noord Holland; L=Amsterdam; O=ME; OU=IT; CN=DockerMirrorBox Intermediate IA d9add2a9f015 2019.08.09 13:09
*  SSL certificate verify ok.
> GET / HTTP/1.1
> Host: k8s.gcr.io
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 502 Bad Gateway
< Server: nginx/1.14.0
< Date: Fri, 09 Aug 2019 13:25:38 GMT
< Content-Type: text/html
< Content-Length: 173
< Connection: keep-alive
<
<html>
<head><title>502 Bad Gateway</title></head>
<body bgcolor="white">
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx/1.14.0</center>
</body>
</html>
* Connection #0 to host 192.168.178.26 left intact
```

**Comment by Hermain** [comment]
And progress here?

**Comment by rpardini** [comment]
I've updated to the latest nginx in `0.3.0-beta1`, could you try that and report back? Thanks

**Comment by zedtux** [comment]
I will check to test this again, but I've switch to Linux since (thanks god, so much faster).

**Comment by rpardini** [comment]
I hit something similar, related to TLS1.2+ only servers and and the old nginx/TLS client version, which has been fixed in 0.3.0.
